### ISSUE TITLE
JDK21 sanity.openjdk failures on x86-64_linux on some machines

### ISSUE THREAD

**Issue by smlambert** [adoptium/aqa-tests#5012]
Some test cases from jdk21 sanity.openjdk targets are failing on x86-64_linux on certain machines.  The failures are in the jdk_lang, jdk_util and jdk_foreign test targets, details can be found in the Jan CPU JDK21 AQA triage, see https://github.com/adoptium/aqa-tests/issues/4983#issuecomment-1896454024

---

**sanity.openjdk** - 5 targets fail, jdk_lang_0, jdk_lang_1, jdk_util_0, jdk_util_1 and jdk_foreign_0 - rerun in [Grinder/8539](https://ci.adoptium.net/job/Grinder/8539) fail, rerun on test-docker-ubuntu2204-x64-1 rerun in [Grinder/8557](https://ci.adoptium.net/job/Grinder/8557) passes 
[Grinder_20240118175328_jdk21_x64Linux.tap.txt](https://github.com/adoptium/aqa-tests/files/13980271/Grinder_20240118175328_jdk21_x64Linux.tap.txt)

---

Marking https://ci.adoptium.net/job/Test_openjdk21_hs_sanity.openjdk_x86-64_linux/151/ as a keep forever to show a good passing run on test-docker-fedora35-x64-1

Failing in https://ci.adoptium.net/job/Test_openjdk21_hs_sanity.openjdk_x86-64_linux/153/ with lots of testcases timing out on test-docker-fedora37-x64-3, with lots of `03:32:59  ACTION: testng -- Failed. Unexpected exit from test [exit code: 137]` and `Error. Agent communication error: java.net.SocketException: Broken pipe; check console log for any additional details` issues

Failing in https://ci.adoptium.net/job/Test_openjdk21_hs_sanity.openjdk_x86-64_linux/155/ on test-docker-fedora35-x64-1 with same types of issues as from the Test_openjdk21_hs_sanity.openjdk_x86-64_linux/153 above.

List of failing test cases from Test_openjdk21_hs_sanity.openjdk_x86-64_linux/155:

```
		java/lang/Package/bootclasspath/GetPackageFromBootClassPath.java.GetPackageFromBootClassPath
		
		java/lang/StackWalker/LocalsAndOperands.java#id1.LocalsAndOperands_id1
		
		java/lang/StrictMath/SqrtTests.java.SqrtTests
		
		java/lang/String/concat/ImplicitStringConcatManyLongs.java.ImplicitStringConcatManyLongs
		
		java/lang/Thread/virtual/stress/Skynet.java#default.Skynet_default
		
		jdk/internal/vm/Continuation/BasicExt.java#COMP_WINDOW_LENGTH_2.BasicExt_COMP_WINDOW_LENGTH_2
		
		java/util/Arrays/TimSortStackSize2.java.TimSortStackSize2
		
		java/util/LinkedHashMap/Basic.java.Basic
		
		java/util/LinkedHashSet/Basic.java.Basic
		
		java/util/List/NestedSubList.java.NestedSubList
		
		java/util/Locale/Bug4152725.java.Bug4152725
		
		java/util/Locale/Bug6989440.java.Bug6989440
		
		java/util/Locale/bcp47u/CurrencyTests.java.CurrencyTests
		
		java/util/StringJoiner/StringJoinerTest.java.StringJoinerTest
		
		java/util/concurrent/tck/JSR166TestCase.java#others.JSR166TestCase_others
		
		java/util/jar/JarFile/jarVerification/MultiProviderTest.java.MultiProviderTest
		
		java/util/jar/JarFile/mrjar/MultiReleaseJarHttpProperties.java.MultiReleaseJarHttpProperties
		
		java/util/jar/JarFile/mrjar/MultiReleaseJarProperties.java.MultiReleaseJarProperties
		
		java/util/jar/Manifest/ValueUtf8Coding.java.ValueUtf8Coding
		
		java/lang/StackWalker/LocalsAndOperands.java#id0.LocalsAndOperands_id0
		
		java/lang/String/concat/ImplicitStringConcatShapes.java.ImplicitStringConcatShapes
		
		java/util/StringJoiner/MergeTest.java.MergeTest
		
		java/util/concurrent/forkjoin/AsyncShutdownNowInvokeAnyRace.java.AsyncShutdownNowInvokeAnyRace
		
		java/util/concurrent/forkjoin/Integrate.java.Integrate
		
		java/util/stream/test/org/openjdk/tests/java/util/stream/mapMultiOpTest.java.mapMultiOpTest
		
		java/util/zip/ZipFile/TestZipFileEncodings.java.TestZipFileEncodings
		
		java/foreign/TestLargeSegmentCopy.java.TestLargeSegmentCopy
		
		java/lang/String/concat/ImplicitStringConcatOOME.java.ImplicitStringConcatOOME
		
		java/util/BitSet/stream/BitSetStreamTest.java.BitSetStreamTest
		
		java/util/HashMap/WhiteBoxResizeTest.java.WhiteBoxResizeTest
		
		java/util/HexFormat/HexFormatTest.java.HexFormatTest

```

**Comment by smlambert** [comment]
[Deep history](https://trss.adoptium.net/deepHistory?testId=657e9f0520748f006f9ffb8a) view looks like:

<img width="1198" alt="Screenshot 2024-01-29 at 2 08 28 PM" src="https://github.com/adoptium/aqa-tests/assets/2836948/a0ed4c8d-47b0-40fe-a727-7dc19d74f29a">

indicating that something happened between Dec 16 and 23 (either in test material or machine configuration to have this issue arise).

**Comment by adamfarley** [comment]
Ok, the following unit tests seem to fail uniquely on the new machine (test-docker-fedora39-x64-1), but passed when run on other fedora machines in the past ([example1](https://ci.adoptium.net/job/Test_openjdk21_hs_sanity.openjdk_x86-64_linux/155/console), [example2](https://ci.adoptium.net/job/Test_openjdk21_hs_sanity.openjdk_x86-64_linux/153/consoleFull)).

sun/security/krb5/MicroTime
java/util/Locale/bug4122700
java/util/Map/InPlaceOpsCollisions
java/util/ResourceBundle/Bug6355009
java/util/Scanner/ScanTest

The [Grinder re-run Stewart launched](https://ci.adoptium.net/job/Grinder/8673) tells us that all of those were infrequent failures, as they passed in the rerun.

In short, I see no consistent failures in sanity.openjdk that have not occurred on existing Fedora machines in the past month, so there's nothing uniquely wrong with this new Fedora machine.

**Comment by smlambert** [comment]
@adamfarley - please continue to investigate this issue to understand what has changed between Dec 16 and 23 that introduces these failures.
- are the failing tests new or changed?
- has there been a change to the machine configuration or the underlying docker host machine?
and so forth

**Comment by adamfarley** [comment]
Will do.

**Comment by adamfarley** [comment]
**Summary**

I think this could be a memory issue caused by a massive concurrency spike.

**Details**

Ok, I've taken a look at the failing test targets, and I'm seeing a 2-3x increase in the runtime on Fedora, along with the most consistent "failed" test targets on Fedora as well. Ubuntu and centos seem to pass the test targets at least some of the time, so I'm concentrating on Fedora for the first pass.

Since 137 is the Linux code for processes being killed due to using up too much memory, this may also explain the socket error if, hypothetically, the VM we're trying to "get" has also been killed.

Still, that's a hypothesis, not proof, so let's see what changed during the Dec 16 and 23 period that could affect memory usage and/or networking.

The first thing I noticed was that the [failures have a concurrency rating of 25](https://trss.adoptium.net/output/test?id=65a8912596a988006d3ba35b) and the [passes have a concurrency rating of 3](https://trss.adoptium.net/output/test?id=657e9f0520748f006f9ffba0). This would explain the timeouts "getting" a VM, and also the memory problems.

Will [kick off some tests with a reduced concurrency](). If that solves the issue, let's tweak the code to make sure this doesn't happen again.

**Comment by adamfarley** [comment]
[Potential candidate for concurrency spike trigger.](https://github.com/adoptium/aqa-tests/commit/e37ba108a1aaab7722e61a1cd384670fa1dde5d5)

**Comment by adamfarley** [comment]
Ok, there appears to be a bug preventing test pipeline jobs from running anything from personal forks, so here's a grinder:

https://ci.adoptium.net/job/Grinder/8712/console

**Comment by adamfarley** [comment]
The grinder successfully reduces the number of concurrent threads in the test. Now to see if reducing the concurrency reduces the number of failures. Leaving to run.

**Comment by sophia-guo** [comment]
Another note is [Deep history](https://trss.adoptium.net/deepHistory?testId=657e9f0520748f006f9ffb8a) shows it mainly failed on the docker agents.
<img width="1057" alt="Screenshot 2024-02-05 at 12 01 23 PM" src="https://github.com/adoptium/aqa-tests/assets/18030625/9fc3b03a-8f9b-4835-a149-03035fdab21b">

Jan 30th, 2024 run on test-ibmcloud-ubuntu1604-x64-1 succeeds. The PR https://github.com/adoptium/aqa-tests/pull/5035 might also be a part of fix , which reduces the `NPROCS` and hence `if $(CORE) < $(MEM)` we won't hit the issue. Only ` if $(CORE) > $(MEM)` the MEM calculation might be an issue.

**Comment by sxa** [comment]
> [Potential candidate for concurrency spike trigger.](https://github.com/adoptium/aqa-tests/commit/e37ba108a1aaab7722e61a1cd384670fa1dde5d5)

Seems likely. I /think/ the correct code should be:
```
else CGMEM=`expr $${KMEMMB} \* 1048576`; fi; CGMEMMB=`expr $${CGMEM} / 1048576`;`
```
So that original PR wasn't correct for all situations

**Comment by adamfarley** [comment]
Ok, the concurrency level should now be within the bounds of sanity. Will kick some jobs off.

**Comment by smlambert** [comment]
Closed via #5063
